README.md
# PySXB
Interface for the WDC development boards

## Installation

OS X & Linux:

    pip install battlenet_client

<!-- ## Clone
Clone the latest version: https://gitlab.com/battlenet1/battlenet-client.git -->

## Usage Example
    from pysxb import PySXB
    

For more information, please the [wiki][wiki]

Note: This package is a dependency of my other [Battle.net REST API packages](https://gitlab.com/battlenet1)

## Release History
Please change log for complete history

## Meta

David "Gahd" Couples – [@gahdania][twitter] – gahdania@gahd.io

Distributed under the GPL v3+ license. See ``LICENSE`` for more information.

[Battle.net Client on Gitlab][gitlab]

## Contributing

1. [Fork it][fork]
2. Create your feature branch (`git checkout -b feature/fooBar`)
3. Commit your changes (`git commit -am 'Add some fooBar'`)
4. Push to the branch (`git push origin feature/fooBar`)
5. Create a new Pull Request

<!-- Markdown link & img dfn's -->
[wiki]: https://battlenet1.gitlab.io/battlenet-client
[twitter]: https://twitter.com/gahdania
[gitlab]: https://gitlab.com/battlenet1/battlenet-client
[fork]: https://gitlab.com/battlenet1/battlenet-client/-/forks/new
[header]: https://gilab.com/

CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.0] - 2021-08-15
### Added
constants.py - added mnemonic constants for the games, and regions
client.py - added str and repr methods

### Changed
util.py - allow localize to except a None type to retrieve all of the locales for an API endpoint
client.py - rewrite of endpoint
          - updated changes for constants
exceptions.py - added BNetNegativeIndexError, and BNetNoDataFoundError

### Removed
None

## [1.1.0] - 2021-07-15
### Added
exceptions.py - for exceptions

### Changed
client.py = _request renamed to endpoint

## [0.0.3] - 2021-06-20
### Added
util.py - for utility functions
tests for the client (need to work on the authorization workflow tests)

### Changed
Moved slugify, localize, currency_convertor into their own file (util.py)
Made modifications class constructor:
  moved standard and condensed into it
  moved client_id and client_secret parameters to positional
  added "release" attribute to allow re-releases of games (ie WoW Classic and TBC as of 2021-Jun
Made modifications to _request:
  added logic for handling locale properly
  added logic to add the namespace parameter to the headers
Removed the "locale" check from validate_token and get_user_info
changed _request to request

### Removed
N/A


## [0.0.2] - 2020-03-25
### Added
BattleNetBaseClass definition
stream keyword for request()


### Changed
Moved get_user_info() and validate_token() to BattleNetBaseClass
corrected some bugs that was preventing BattleNetUserAuthClient from
    successfully connecting to the API

### Removed
N/A

## [0.0.1] - 2019-12-08
### Added
Initial layout
BattleNetCredentialClient class
BattleNetUserAuthClient class
localize() function to standardize the locale parameters

### Changed
N/A

### Removed
N/A

[0.0.3]: https://gitlab.com/battlenet/tags/v0.0.3
[0.0.2]: https://gitlab.com/battlenet/tags/v0.0.2


main.py
import getopt
import sys
from serial import SerialTimeoutException

from PySXB.pysxb import PySXB


if __name__ == "__main__":

    argv = sys.argv[1:]
    device = None
    baud = None
    file = None
    cpu_mode = None
    exec_address = None
    opts = []
    short_opts = "d:b:f:e:E:"
    long_opts = ("device =", "baud =", "file =", "C02 =", "C816 =")
    try:
        opts, opt_args = getopt.getopt(argv, short_opts, long_opts)
    except getopt.GetoptError as opt_error:
        print(f"Error: {opt_error}\n")

    for opt, opt_arg in opts:

        if opt in ('-d', '--device'):
            device = opt_arg

        if opt in ('-b', '--baud'):
            baud = opt_arg

        if opt in ('-f', '--file'):
            file = opt_arg

        if opt in ('-e', '--C02'):
            cpu_mode = 2       #
            exec_address = opt_arg

        if opt in ('-E', '--C816'):
            cpu_mode = 1
            exec_address = opt_arg

    # set defaults for needed variables
    if not device:
        device = '/dev/ttyUSB0'

    if not baud:
        baud = 9600

    if not cpu_mode:
        cpu_mode = 1

    ser = PySXB(device, baud)

    if file:
        ser.load_program(file)

    try:
        ser.execute(cpu_mode, 0x2000, 2)       # equivalent to F9 in debugger w/o break points
    except KeyboardInterrupt:
        sys.exit()
    except SerialTimeoutException as error:
        print(f"Error: {error}")
        sys.exit(1)
    else:
        print("End!")


baird.py
# This is a sample Python script.
# A basic implimentation of the Terbium IDE interface protocol
# used with the WDC W65C816SXB evalutation board, so that Unix
# users don't have to be anywhere near that Windows crapware.
# --
# Chris Baird,, <cjb@brushtail.apana.org.au> July 2019# Licensed as per the GPL-3, for what it's worth.

import time
import serial
import sys

ser = serial.Serial(
    port='/dev/ttyUSB0',
    baudrate=9600,
    parity=serial.PARITY_NONE,
    stopbits=serial.STOPBITS_ONE,
    timeout=10,
    write_timeout=10,
    bytesize=serial.EIGHTBITS)


def startcmd(cmd):
    # ser.write(chr(85)+chr(170))
    ser.write((85, 170))
    r = ord(ser.read(1))
    if r != 204:
        print("Response error = ", r)
        sys.exit(1)

    # ser.write(chr(cmd))
    ser.write(cmd)


def readmem(addr, length):
    startcmd(3)
    addrba = (addr >> 16) & 255
    addrhi = (addr >> 8) & 255
    addrlo = addr & 255
    lenlo = length & 255
    lenhi = (length >> 8) & 255
    # ser.write(chr(addrlo)+chr(addrhi)+chr(addrba)+chr(lenlo)+chr(lenhi))
    ser.write((addrlo, addrhi, addrba, lenlo, lenhi))

    # result = []
    result = bytearray(ser.read(length))
    # for _ in range(length):
    #     result.append(ord(ser.read(1)))

    return result


def pokemem(addr, val):
    startcmd(2)
    addrba = addr >> 16
    addrlo = addr & 255
    addrhi = (addr & 65280) >> 8
    # ser.write(chr(addrlo)+chr(addrhi)+chr(addrba)+chr(1)+chr(0))
    command = (addrlo, addrhi, addrba, 0x01, 0x00)
    ser.write(command)
    ser.write(val)

    r = ord(ser.read(1))
    if r != 0xcc:
        return False

    return True


def writemem(addr, vallist):
    startcmd(2)
    addrba = (addr >> 16) & 255
    addrhi = (addr >> 8) & 255
    addrlo = addr & 255
    lenlo = len(vallist) & 255
    lenhi = (len(vallist) >> 8) & 255

    # ser.write(chr(addrlo)+chr(addrhi)+chr(addrba)+chr(lenlo)+chr(lenhi))
    ser.write((addrlo, addrhi, addrba, lenlo, lenhi))

    # for i in range(len(vallist)):
    #     # NetBSD has non-existant userspace delays, hence this kludge
    #     if len(vallist) > 128:
    #         for t in range(10000):
    #             a = t
    #     ser.write(chr(vallist[i]))
    for index in vallist:
        ser.write(index)

    r = ord(ser.read(1))
    if r != 0xcc:
        return False

    return True


def getargvals(argstr):
    if len(argstr) == 1:
        return int(argstr)

    if (argstr[0] == '$') or (argstr[0] == 'x') or (argstr[0] == '0'):
        result = int(argstr[1:], 16)
    else:
        result = int(argstr)
    return result


sys.argv.pop(0)
dumpstart = -1
dumpend = -1
s = sys.argv[0]
DUMPSIZE = 16

if s == "-r":
    dumpstart = getargvals(sys.argv[1])
    dumpend = getargvals(sys.argv[2])
    if (dumpstart > -1) and (dumpstart < dumpend):
        dump = readmem(dumpstart, dumpend-dumpstart+1)
        while len(dump):
            l = "%04x: " % (dumpstart & 65535)
            s = ""
            a = dump[:DUMPSIZE]
            dumpstart += len(a)
            for i in a:
                l += "%02x " % i
                c = i & 127
                if c < 32:
                    c += 64
                s += chr(c)
            print(l.ljust(54), s)
            dump = dump[DUMPSIZE:]

#

if s == "-b":
    dumpstart = getargvals(sys.argv[1])
    dumpend = getargvals(sys.argv[2])
    if (dumpstart > -1) and (dumpstart < dumpend):
        dump = readmem(dumpstart, dumpend-dumpstart+1)
        for i in range(dumpend-dumpstart+1):
            sys.stdout.write(chr(dump[i]))


if s == "-s":
    l = []
    br = open(sys.argv[1], "rb").read()
    for b in br:
        l.append(ord(b))
    writemem(getargvals(sys.argv[2]), l)

#

BLK = 128
if s == "-S":
    a = getargvals(sys.argv[2])
    # l = []
    br = open(sys.argv[1], "rb").read()
    # for b in br:
    #     l.append(hex(b))

    # for ad in range(a, a+len(l), BLK):
    #     co = l[:BLK]
    #     print(hex(ad), chr(13),)
    #     sys.stdout.flush()
    #     l = l[BLK:]
    #
    start_blk = 0
    p = 0
    for ad in range(a, a+len(br), BLK):

        while p != -1:
            # writemem(ad, co)
            # mem = readmem(ad, len(co))
            # if mem == co:
            #     break
            if not writemem(ad, br[start_blk:start_blk+BLK]):  #
                if p == 10:
                    print(f"{ad:#04x}:  giving up\n")
                    sys.exit(-10)

                else:
                    p += 1          # p = p + 1
                    print(f"Redoing address: {ad:#04x}\n")
                    print(f"{ad:#04x}: retry = {p:02d}\n")
            else:
                start_blk += BLK


if (s == "-e") or (s == "-E"):
    if s == "-E":
        cpumode = 0             # 65816 mode
    else:
        cpumode = 1             # 6502 mode

    execaddr = getargvals(sys.argv[1])-1  # "RTI trick", hence the -1
    execlo = execaddr & 255
    exechi = (execaddr >> 8) & 255

    procstate = [0, 0,           # 0,1 .A
                 0, 0,           # 2,3 .X
                 0, 0,           # 4,5 .Y
                 execlo, exechi, # 6,7 exec addr
                 0,              # 8 direct page register
                 0,              # 9 ?processor status bits
                 255,            # A stack pointer
                 0,              # B ?processor status bits
                 0,              # C
                 cpumode,        # D CPU mode (0=65816, 1=6502)
                 0,              # E ?.B
                 0]              # F
    writemem(32256, procstate)
    startcmd(5)


src/PySXB/pysxb.py
"""PySXB:  Basic python implementation of Terbium IDE, (TIDE).

Based on the work of Chris Baird <cjb@brushtail.apana.org.au> July 2019

David Couples
2021 November
"""

import serial

AT = (0x55, 0xAA)
# ???  = 0x01
WRITE = 0x02
READ = 0x03
TIDE = 0x04
EXEC = 0x05
BLK_SIZE = 62
OK = 0xCC


class PySXB(serial.Serial):
    """

    Args:
        port (str): the device path for the serial port of the SXB board
        speed (int): speed of communication (baud rate) default 9600
        emulation_mode (bool): Sets the if emulation is turned on or off
            defaults to True.  W65C02SXB also should be true
        *args: additional positional args to pass onto parent class
        **kwargs: additional keyword args to pass onto parent class
    """

    def __init__(self, port, emulation_mode=True, *args, **kwargs):
        super().__init__(port, 9600, *args, **kwargs)

        # load TIDE data from the board
        if self._command(TIDE):
            tide_data = bytearray(self.read(0x20))
            self.mon_ram = self.decode(tide_data[:3])
            self.cpu_type = tide_data[3]
            self.board_id = tide_data[4]
            self.mon_rom = self.decode(tide_data[8:11])
            self.shadow_vector_base = self.decode(tide_data[11:14])
            self.hw_io = self.decode(tide_data[14:17])
            self.hw_vector_base = self.decode(tide_data[17:20])

            if not self.cpu_type:
                self.emulation = True

            if self.cpu_type:
                self.emulation = emulation_mode

    @staticmethod
    def decode(byte_array):
        return_val = 0
        for i, byte in enumerate(byte_array):
            return_val |= byte << i * 8

        return return_val

    @staticmethod
    def encode(number, address=True):
        if address:
            return (number & 255), ((number >> 8) & 255), ((number >> 16) & 255)
        
        return (number & 255), ((number >> 8) & 255)

    def hex(self, data, start_address=None):
        line = ""

        if not start_address:
            start_address = 0

        if isinstance(data, int):
            data = self.encode(data, False)

        for i, datum in enumerate(data):

            if i == 0:
                line = f"{start_address:#06x}:"
            if i > 0 and i % 16 == 0:
                line += f"\n{i+start_address:#06x}:"
            if (i % 8 == 0) and (i % 16 != 0):
                line += " "
            line += f" {datum:02X}"

        return line

    @property
    def stack(self):
        return self.read_board(0x100, 256)

    @property
    def processor(self):
        return self.read_board(self.mon_ram, 16)

    @property
    def zero_page(self):
        return self.read_board(0x00, 256)

    @property
    def irq_vector(self):
        address = self.shadow_vector_base + 10

        if self.emulation:
            return address + 16

        return address

    @property
    def reset_vector(self):
        address = self.shadow_vector_base + 8

        if self.emulation:
            return address + 16

        return address

    @property
    def nmi_vector(self):
        address = self.shadow_vector_base + 6

        if self.emulation:
            return address + 16

    @property
    def break_vector(self):
        address = self.shadow_vector_base + 2

        if self.emulation:
            return address + 24

        return address

    @property
    def abort_vector(self):
        address = self.shadow_vector_base + 4

        if not self.cpu_type:
            return None

        if self.emulation:
            return address + 16

        return address

    @property
    def coprocessor_vector(self):
        address = self.shadow_vector_base

        if not self.cpu_type:
            return None

        if self.emulation:
            return address + 16

        return address

    def hardware_address(self, address):
        return self.hw_vector_base & 0xff00 | address & 0x00ff

    def _at(self):
        if self.write(AT):
            try:
                return ord(self.read(1)) == OK
            except ValueError:
                return False

    def _command(self, cmd_code, *, address=None, length=None):

        if not self._at():
            return None

        command = bytearray((cmd_code,))

        if address:
            for element in self.encode(address):
                command.append(element)

        if length:
            for element in self.encode(length, False):
                command.append(element)

        try:
            return self.write(command) > 0
        except serial.SerialTimeoutException as error:
            print(f"Error: {error}\n")
            return None
        except TypeError as error:
            print(f"Error: {error}\n")
            return None

    def write_board(self, address, data):
        length = len(data)

        if self._command(WRITE, address=address, length=length):
            if length < BLK_SIZE:
                return self.write(data)

            for blk in range(0, length, BLK_SIZE):
                if blk + BLK_SIZE < length:
                    self.write(data[blk:blk+BLK_SIZE])
                    continue
                self.write(data[blk:length])

        return 0

    def execute(self, address, a_reg=0, x_reg=0, y_reg=0, processor_flags=0x76):
        """
        """
        a_reg = self.encode(a_reg, False)
        x_reg = self.encode(x_reg, False)
        y_reg = self.encode(y_reg, False)

        address = self.encode(address)
        processor_state = (
            a_reg,                      # A Register
            x_reg,                      # X register
            y_reg,                      # Y register
            address[0], address[1],     # Program Counter
            0, 0,                       # Direct Register
            255, 1,                     # Stack Pointer
            processor_flags,            # Processor status flags
            self.emulation,             # Emulation Off/On
            0,                          # Program Bank Register
            0                           # Data Bank Register
        )

        self.write_board(self.mon_ram, processor_state)
        return self._command(EXEC, length=1)

    def read_board(self, address, length):

        self._command(READ, address=address, length=length)
        if length < BLK_SIZE:
            return self.read(length)

        temp = bytearray(length)
        for blk in range(0, length, BLK_SIZE):
            if blk + BLK_SIZE < length:
                temp[blk: blk + BLK_SIZE] = self.read(BLK_SIZE)
                continue
            temp[blk:length] = self.read(length-blk)
        return temp

    def load(self, file):
        with open(file, 'rb') as fh:
            rom_data = fh.read()

        if rom_data[0] == 0x5a:     # checking for SXB flag symbol from the WDC assembler/Compiler

            code_address = self.decode(rom_data[1:4])
            code_length = self.decode(rom_data[4:6])
            code = rom_data[7:7+code_length]
            self.write_board(code_address, code)
            return code_address, code_length

        raise ValueError("Program needs to be compiled/assembled with the -g option")
